File: frontend/src/lib/logParser.ts
import { StreamMessage, RunMetadata } from './types';

/**
 * Parses Claude Code stream-json output into human-readable terminal lines.
 */

export interface TerminalLine {
    text: string;
    style: "text" | "tool" | "tool-result" | "system" | "result" | "error" | "dim";
}

/**
 * Parse a single log line (may be JSON or plain text) into
 * one or more formatted terminal lines.
 */
export function parseLogLine(raw: string): TerminalLine[] {
    const trimmed = raw.trim();
    if (!trimmed) return [];

    // Try to parse as JSON
    let json: Record<string, unknown>;
    try {
        json = JSON.parse(trimmed);
    } catch {
        // Not JSON — show as plain text
        return [{ text: trimmed, style: "text" }];
    }

    const type = json.type as string | undefined;

    // assistant message (contains text + tool_use blocks)
    if (type === "assistant" && json.message) {
        return parseAssistantMessage(json.message as Record<string, unknown>);
    }

    // stream_event (token-level delta)
    if (type === "stream_event" && json.event) {
        return parseStreamEvent(json.event as Record<string, unknown>);
    }

    // user message (contains tool_result blocks after tool calls)
    if (type === "user" && json.message) {
        return parseUserMessage(json.message as Record<string, unknown>);
    }

    // system messages — only show actionable ones, skip noise
    if (type === "system") {
        const subtype = json.subtype as string | undefined;
        if (subtype === "api_retry") {
            const attempt = json.attempt ?? "?";
            const error = json.error ?? "";
            return [{ text: `[Retry ${attempt}] ${error}`, style: "system" }];
        }
        // Skip noisy system messages (init, task_notification, task_progress, etc.)
        return [];
    }

    // Interactive session messages
    if (type === "prompt_start") {
        const promptText = (json.prompt as string) || "";
        const promptNum = (json.prompt_number as number) || 0;
        const lines: TerminalLine[] = [];
        lines.push({ text: `--- Prompt #${promptNum} ---`, style: "system" });
        if (promptText) {
            lines.push({ text: `> ${promptText}`, style: "system" });
        }
        return lines;
    }

    if (type === "idle") {
        return [{ text: "[Bereit]", style: "dim" }];
    }

    if (type === "keepalive") {
        return [];
    }

    // Suppress noisy event types
    if (type === "rate_limit_event" || type === "content_block_start" ||
        type === "content_block_delta" || type === "content_block_stop" ||
        type === "message_start" || type === "message_stop" ||
        type === "ping" || type === "error_event") {
        return [];
    }

    // result message (final output)
    if (type === "result") {
        const lines: TerminalLine[] = [];
        const result = json.result as string | undefined;
        const subtype = json.subtype as string | undefined;
        const cost = json.cost_usd as number | undefined;
        const duration = json.duration_ms as number | undefined;
        const turns = json.num_turns as number | undefined;

        if (result) {
            lines.push({ text: result, style: "result" });
        }

        const meta: string[] = [];
        if (subtype) meta.push(subtype === "success" ? "Erfolgreich" : subtype);
        if (cost != null) meta.push(`$${cost.toFixed(4)}`);
        if (duration != null) meta.push(`${(duration / 1000).toFixed(1)}s`);
        if (turns != null) meta.push(`${turns} Turns`);
        if (meta.length > 0) {
            lines.push({ text: `── ${meta.join(" | ")} ──`, style: "dim" });
        }
        return lines;
    }

    // Unknown JSON — suppress rather than showing raw
    if (type) {
        return [];
    }
    return [{ text: trimmed, style: "text" }];
}

function parseUserMessage(message: Record<string, unknown>): TerminalLine[] {
    const lines: TerminalLine[] = [];
    const content = message.content as Array<Record<string, unknown>> | undefined;
    if (!Array.isArray(content)) return lines;

    for (const block of content) {
        const blockType = block.type as string;
        if (blockType === "tool_result") {
            const resultContent = block.content as string | undefined;
            if (resultContent) {
                const preview = resultContent.length > 300
                    ? resultContent.slice(0, 300) + "..."
                    : resultContent;
                lines.push({ text: preview, style: "tool-result" });
            }
        }
    }
    return lines;
}

function parseAssistantMessage(message: Record<string, unknown>): TerminalLine[] {
    const lines: TerminalLine[] = [];
    const content = message.content as Array<Record<string, unknown>> | undefined;
    if (!Array.isArray(content)) return lines;

    for (const block of content) {
        const blockType = block.type as string;

        if (blockType === "text") {
            const text = block.text as string;
            if (text?.trim()) {
                lines.push({ text: text.trim(), style: "text" });
            }
        } else if (blockType === "tool_use") {
            const name = block.name as string || "Tool";
            const input = block.input as Record<string, unknown> | undefined;
            const desc = formatToolInput(name, input);
            lines.push({ text: `$ ${name} ${desc}`, style: "tool" });
        } else if (blockType === "tool_result") {
            const content = block.content as string | undefined;
            if (content) {
                // Truncate long tool results
                const preview = content.length > 300
                    ? content.slice(0, 300) + "..."
                    : content;
                lines.push({ text: preview, style: "tool-result" });
            }
        }
    }

    return lines;
}

function parseStreamEvent(event: Record<string, unknown>): TerminalLine[] {
    const eventType = event.type as string;

    // content_block_delta — streaming text
    if (eventType === "content_block_delta") {
        const delta = event.delta as Record<string, unknown> | undefined;
        if (delta?.type === "text_delta" && delta.text) {
            // Skip individual deltas — they'll be in the full assistant message
            return [];
        }
    }

    // content_block_start
    if (eventType === "content_block_start") {
        const block = event.content_block as Record<string, unknown> | undefined;
        if (block?.type === "tool_use") {
            const name = block.name as string || "Tool";
            return [{ text: `$ ${name}...`, style: "tool" }];
        }
    }

    // message_start, message_stop — skip
    if (eventType === "message_start" || eventType === "message_stop" ||
        eventType === "content_block_stop" || eventType === "content_block_delta") {
        return [];
    }

    return [];
}

function formatToolInput(name: string, input: Record<string, unknown> | undefined): string {
    if (!input) return "";

    // Common tools with nice formatting
    if (name === "Read" || name === "Glob" || name === "Grep") {
        return (input.file_path || input.path || input.pattern || "") as string;
    }
    if (name === "Edit" || name === "Write") {
        return (input.file_path || "") as string;
    }
    if (name === "Bash") {
        const cmd = (input.command || "") as string;
        return cmd.length > 80 ? cmd.slice(0, 80) + "..." : cmd;
    }
    if (name === "Task" || name === "WebSearch" || name === "WebFetch") {
        return (input.description || input.query || input.url || "") as string;
    }

    // Fallback: show first string value
    for (const val of Object.values(input)) {
        if (typeof val === "string" && val.length > 0) {
            return val.length > 80 ? val.slice(0, 80) + "..." : val;
        }
    }
    return "";
}

/**
 * Parse all raw log lines into formatted terminal lines.
 */
export function parseLogLines(rawLines: string[]): TerminalLine[] {
    const result: TerminalLine[] = [];
    for (const line of rawLines) {
        result.push(...parseLogLine(line));
    }
    return result;
}

/**
 * Extract cost/duration/turns metadata from stream messages.
 */
export function extractRunMetadata(lines: StreamMessage[]): RunMetadata | null {
    for (const line of lines) {
        if (line.type !== "output") continue;
        try {
            const data = JSON.parse(line.data);
            if (data.type === "result") {
                return {
                    cost: data.cost_usd ?? data.result?.cost_usd ?? undefined,
                    duration: data.duration_ms ?? data.result?.duration_ms ?? undefined,
                    turns: data.num_turns ?? data.result?.num_turns ?? undefined,
                };
            }
        } catch {
            continue;
        }
    }
    return null;
}

File: backend/app/log_parser.py
"""
Parses Claude Code stream-json output into human-readable text.
Python port of frontend/src/lib/logParser.ts.
"""

import json


def parse_log_line(raw: str) -> list[str]:
    """Parse a single log line (JSON or plain text) into readable strings."""
    trimmed = raw.strip()
    if not trimmed:
        return []

    try:
        data = json.loads(trimmed)
    except (json.JSONDecodeError, ValueError):
        return [trimmed]

    if not isinstance(data, dict):
        return [trimmed]

    msg_type = data.get("type")

    # assistant message (text + tool_use blocks)
    if msg_type == "assistant" and data.get("message"):
        return _parse_assistant_message(data["message"])

    # stream_event (token-level delta) — skip, full message comes later
    if msg_type == "stream_event":
        return []

    # system messages
    if msg_type == "system":
        subtype = data.get("subtype", "")
        if subtype == "api_retry":
            return [f"[Retry {data.get('attempt', '?')}] {data.get('error', '')}"]
        if subtype == "init":
            return ["Session gestartet"]
        return []

    # result message
    if msg_type == "result":
        lines = []
        result = data.get("result")
        if result:
            lines.append(str(result))
        meta = []
        subtype = data.get("subtype")
        if subtype:
            meta.append("Erfolgreich" if subtype == "success" else subtype)
        cost = data.get("cost_usd")
        if cost is not None:
            meta.append(f"${cost:.4f}")
        duration = data.get("duration_ms")
        if duration is not None:
            meta.append(f"{duration / 1000:.1f}s")
        turns = data.get("num_turns")
        if turns is not None:
            meta.append(f"{turns} Turns")
        if meta:
            lines.append(f"── {' | '.join(meta)} ──")
        return lines

    return []


def _parse_assistant_message(message: dict) -> list[str]:
    lines = []
    content = message.get("content")
    if not isinstance(content, list):
        return lines

    for block in content:
        if not isinstance(block, dict):
            continue
        block_type = block.get("type")

        if block_type == "text":
            text = block.get("text", "").strip()
            if text:
                lines.append(text)

        elif block_type == "tool_use":
            name = block.get("name", "Tool")
            desc = _format_tool_input(name, block.get("input"))
            lines.append(f"$ {name} {desc}")

        elif block_type == "tool_result":
            content_str = block.get("content", "")
            if isinstance(content_str, str) and content_str:
                preview = content_str[:300] + "..." if len(content_str) > 300 else content_str
                lines.append(preview)

    return lines


def _format_tool_input(name: str, input_data: dict | None) -> str:
    if not input_data or not isinstance(input_data, dict):
        return ""

    if name in ("Read", "Glob", "Grep"):
        return str(input_data.get("file_path") or input_data.get("path") or input_data.get("pattern") or "")
    if name in ("Edit", "Write"):
        return str(input_data.get("file_path") or "")
    if name == "Bash":
        cmd = str(input_data.get("command") or "")
        return cmd[:80] + "..." if len(cmd) > 80 else cmd
    if name in ("Task", "WebSearch", "WebFetch"):
        return str(input_data.get("description") or input_data.get("query") or input_data.get("url") or "")

    for val in input_data.values():
        if isinstance(val, str) and val:
            return val[:80] + "..." if len(val) > 80 else val
    return ""


def parse_log_lines(raw_lines: list[str]) -> list[str]:
    """Parse multiple raw log lines into readable text lines."""
    result = []
    for line in raw_lines:
        result.extend(parse_log_line(line))
    return result
